### PR TITLE
Code changes to compare Summary & Description

### PR DESCRIPTION
**Pull request by sarangan12**
ERROR: The summary and description values should not be same.
Path: file:///C:/workspace/autorest/src/modeler/AutoRest.Swagger.Tests/Resource/Swagger/Validation/swagger-summary-description.json#/paths/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/Microsoft.Cache/Redis?api-version={apiVersion}/post
Severity: Error
Category: SDKViolation

Related to https://github.com/Azure/autorest/issues/1826 

No regerate changes

@veronicagg @vishrutshah @dsgouda @salameer Please approve

### CODE REVIEW

File: src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+﻿using AutoRest.Swagger.Validation.Core;
```
Comment (vishrutshah) [inline src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs]: Please add license header to new file

File: src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+using System.Collections.Generic;
```
Comment (vishrutshah) [inline src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs]: Please sort the usings

File: src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        public override string MessageTemplate => Resources.SummaryDescriptionVaidationError;
+
+        /// <summary>
+        /// The severity of this message (ie, debug/info/warning/error/fatal, etc)
+        /// </summary>
+        public override Category Severity => Category.Error;
+
+        public override IEnumerable<ValidationMessage> GetValidationMessages(Operation operation, RuleContext context)
+        {
+            if(operation.Description != null && operation.Summary != null && operation.Description.Trim().Equals(operation.Summary.Trim()))
```
Comment (vishrutshah) [inline src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs]: Should we check for case insensitive string match?

File: src/modeler/AutoRest.Swagger.Tests/Resource/Swagger/Validation/swagger-summary-description.json [hunk 4]
```diff
@@ -0,0 +1,177 @@
+{
```
Comment (vishrutshah) [inline src/modeler/AutoRest.Swagger.Tests/Resource/Swagger/Validation/swagger-summary-description.json]: just curios, could we leverage already existing test file that we are using for other description checks that way we do not explode test files with rules.

File: src/core/AutoRest.Core/Properties/Resources.Designer.cs [hunk 5]
```diff
@@ -137,22 +137,20 @@ public class Resources {
         ///   Looks up a localized string similar to Plugins:
         ///  CSharp:
         ///    TypeName: PluginCs, AutoRest.CSharp
+        ///  JsonRpcClient:
```
Comment (vishrutshah) [inline src/core/AutoRest.Core/Properties/Resources.Designer.cs]: Is this changes intentional?

File: src/core/AutoRest.Core/Properties/Resources.Designer.cs [hunk 6]
```diff
@@ -137,22 +137,20 @@ public class Resources {
...
+        ///    TypeName: PluginCs, AutoRest.CSharp.JsonRpcClient
         ///  Azure.CSharp:
         ///    TypeName: PluginCsa, AutoRest.CSharp.Azure
         ///  Azure.CSharp.Fluent:
         ///    TypeName: PluginCsaf, AutoRest.CSharp.Azure.Fluent
+        ///  Azure.JsonRpcClient:
```
Comment (vishrutshah) [inline src/core/AutoRest.Core/Properties/Resources.Designer.cs]: Is this changes intentional?

File: src/core/AutoRest.Core/Properties/Resources.Designer.cs [hunk 7]
```diff
@@ -674,7 +681,7 @@ public class Resources {
         }
         
         /// <summary>
-        ///   Looks up a localized string similar to For the tracked resource &apos;{0}&apos;, the x-ms-pageable must be same for list by resource group and subscriptions operations..
+        ///   Looks up a localized string similar to For the tracked resource &apos;{0}&apos;, the x-ms-pageable extension values must be same for list by resource group and subscriptions operations..
```
Comment (vishrutshah) [inline src/core/AutoRest.Core/Properties/Resources.Designer.cs]: Is this changes intentional?

File: src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs (hunk 1 above)
Comment (sarangan12) [inline src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs]: Done

File: src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs (hunk 2 above)
Comment (sarangan12) [inline src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs]: Done

File: src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs (hunk 3 above)
Comment (sarangan12) [inline src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs]: Done

File: src/modeler/AutoRest.Swagger.Tests/Resource/Swagger/Validation/swagger-summary-description.json (hunk 4 above)
Comment (sarangan12) [inline src/modeler/AutoRest.Swagger.Tests/Resource/Swagger/Validation/swagger-summary-description.json]: Concerned about more test cases?? :-) I believe every scenario must have its own test case to avoid confusions later

File: src/core/AutoRest.Core/Properties/Resources.Designer.cs (hunk 5 above)
Comment (sarangan12) [inline src/core/AutoRest.Core/Properties/Resources.Designer.cs]: Not related to my change. But since this file is autogenerated I suspect this has something to do with latest change in the master branch

File: src/core/AutoRest.Core/Properties/Resources.Designer.cs (hunk 6 above)
Comment (sarangan12) [inline src/core/AutoRest.Core/Properties/Resources.Designer.cs]: Not related to my change. But since this file is autogenerated I suspect this has something to do with latest change in the master branch

File: src/core/AutoRest.Core/Properties/Resources.Designer.cs (hunk 7 above)
Comment (sarangan12) [inline src/core/AutoRest.Core/Properties/Resources.Designer.cs]: Yes. Related to a recent comment from Veronica

**Comment by sarangan12** [comment]
{
  "type": "Error",
  "code": "SummaryDescriptionValidation",
  "message": "The summary and description values should not be same.",
  "id": "M2023",
  "validationCategory": "SDKViolation",
  "providerNamespace": null,
  "resourceType": null,
  "sources": [
    "file:///C:/workspace/autorest/src/modeler/AutoRest.Swagger.Tests/Resource/Swagger/Validation/swagger-summary-description.json:54:6 ($.paths[\"/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/Microsoft.Cache/Redis?api-version={apiVersion}\"].post)"
  ]
}

File: src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs [hunk 8]
```diff
@@ -33,7 +37,7 @@ public class SummaryDescriptionValidation: TypedRule<Operation>
 
         public override IEnumerable<ValidationMessage> GetValidationMessages(Operation operation, RuleContext context)
         {
-            if(operation.Description != null && operation.Summary != null && operation.Description.Trim().Equals(operation.Summary.Trim()))
+            if(operation.Description != null && operation.Summary != null && operation.Description.Trim().Equals(operation.Summary.Trim(), StringComparison.CurrentCultureIgnoreCase))
```
Comment (vishrutshah) [inline src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs]: Why did you decide to use culture aware case-insensitive comparison by `CurrentCultureIgnoreCase`.

We should use do `case-insensitive` + `culture-insensitive` comparison by using `StringComparison.OrdinalIgnoreCase`

File: src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs [hunk 9]
```diff
@@ -37,7 +37,7 @@ public class SummaryDescriptionValidation: TypedRule<Operation>
 
         public override IEnumerable<ValidationMessage> GetValidationMessages(Operation operation, RuleContext context)
         {
-            if(operation.Description != null && operation.Summary != null && operation.Description.Trim().Equals(operation.Summary.Trim(), StringComparison.CurrentCultureIgnoreCase))
+            if(operation.Description != null && operation.Summary != null && operation.Description.Trim().Equals(operation.Summary.Trim(), StringComparison.OrdinalIgnoreCase))
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs]: Use null coalesce, makes it cleaner and easier to read.

File: src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        /// <summary>
+        /// The severity of this message (ie, debug/info/warning/error/fatal, etc)
+        /// </summary>
+        public override Category Severity => Category.Error;
+
+        public override IEnumerable<ValidationMessage> GetValidationMessages(Operation operation, RuleContext context)
+        {
+            if(operation.Description != null && operation.Summary != null && operation.Description.Trim().Equals(operation.Summary.Trim(), StringComparison.OrdinalIgnoreCase))
+            {
+                yield return new ValidationMessage(new FileObjectPath(context.File, context.Path), this, new object[0]);
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs]: DO take into consideration the path this rule should be reporting.

Comment (sarangan12) [inline src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs]: @dsgouda What does the comment mean?

Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs]: When you return FileObjectPath, the path where this violation occurs would point to the Operation node in this case, if you want it to report as the description node or the summary node you may have to do an AppendProperty like here https://github.com/Azure/autorest/blob/e1729aeda9c4ade2bb07dd8314707f4c1f5e5f00/src/modeler/AutoRest.Swagger/Validation/ListOperationNamingWarning.cs#L46

File: src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs (hunk 9 above)
Comment (sarangan12) [inline src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs]: null coalesce are used to return the values. Are you suggesting them to use here? I am not sure how much it will improve readability. Not changing now

File: src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs (hunk 10 above)
Comment (sarangan12) [inline src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs]: I see what you mean. In this case I actually want to point to the operation node and not the summary or description nodes.

File: src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs (hunk 9 above)
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs]: You could do:
```
operation.Description?.Trim().Equals(operation.Summary?.Trim(), StringComparison.OrdinalIgnoreCase)==true
```

Comment (sarangan12) [inline src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs]: I guessed so.. :-) You have just pointed me the one of the simplest pattern I hate most in C#..... :-)

Comment (balajikris) [inline src/modeler/AutoRest.Swagger/Validation/SummaryDescriptionValidation.cs]: I think you meant `null-conditional  --> ?.` . Agreed that it makes it comparing against a bool harder, as the result of the expression would be a `Nullable<bool>`.
